### PR TITLE
Annualized data support

### LINKED ISSUE #80: Add support for annualized data.
Corresponding backend ticket is https://github.com/Recidiviz/recidiviz-data/issues/6644

**PRD:** [Bookmark](https://docs.google.com/document/d/1eOiMzk8_eDjC4W0xn89qlkmj_6SOqi1dBTyEusAGGaI/edit?ts=5fb45edc#bookmark=id.wgq2v7cyggo7)
**Mocks:** [Figma](https://www.figma.com/file/66HyjH0efZMoAuTlhnojWh/Justice-Counts-Report?node-id=1130%3A19)

Currently, we only show data that is either "point-in-time" (like populations), or aggregated monthly. However, this excludes a lot of the **yearly-aggregated data** that we actually store in the database, giving the user a sense that there is less data available publicly than there actually is. At the same time, we want to prioritize showing monthly data where we can. 

Our solution for this will be to implement an "annual" switch that allows users to see all metrics aggregated to an annual statistic. This switch should exist at the top of the page; switching it should result in the following changes. 

Group 70 (1)

## Metric Blocks 
- A percent change should still show up on these metric blocks. 
- The date range shown on the block should now correspond to the start and end dates for the aggregated data. For example, if the data for Parole Revocations was aggregated between July 2018 and June 2019, that is the date range that should appear on the block.
- For any population counts (Prison, Probation, Parole), use the population from the last month in the date range of interest (whatever is used in the flow diagram). 

## Flow Diagram 
- The primary date range should correspond with the aggregation range for metrics that are annualized in the original report where possible. 
    - If there are multiple metrics that are originally reported on an annual time range but they have conflicting time ranges, the primary date range should align with the majority.
- For any population counts (Prison, Probation, Parole), use the population from the last month in the date range of interest. 
- In some cases, monthly data will be artificially restricted in order to better match data that is only reported annually. In this case, we should indicate this on the card by including a warning tooltip with the following copy and design:
> Data is restricted to older reports in order to match the other reports shown; however, more recent data is available (as of January 2021). Use the Reporting Frequency switch at the top of the page to see the most recent monthly data. 

Group 71 (3)

As a bonus, the orange link in the design above ("Reporting Frequency switch") should scroll the user back to the page position where the data reporting frequency link can be found.

## Graphs
- If possible, bar graphs should be used instead of line graphs for annualized data. 
- Metrics should still appear on the X-axis aligned with a month/year. The month/year should be the last month/year within the aggregation range (e.g., if a particular data point was generated from event counts between January 2019 and December 2019, the date for that metric should be 12/2019).
- The “1 year” option should **not** be present for annual data. Thus, the only time range options for annual data should be “5 years” and “All Time”.

Group 65 (1)

### PR DESCRIPTION
**Pull request by nasaownsky**
## Description of the change

1. Added annualized section to `MainPage` along with new `Switch` component
2. Updated chart logic

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Configuration change (adjusts configuration to achieve some end related to functionality, development, performance, or security)

## Related issues

> Closes #80 

### Development

These boxes should be checked by the submitter prior to merging:

- [x] Manual testing against realistic data has been performed locally

### Code review

These boxes should be checked by reviewers prior to merging:

- [x] This pull request has a descriptive title and information useful to a reviewer
- [ ] This pull request has been moved out of a Draft state, has no "Work In Progress" label, and has assigned reviewers
- [x] Potential security implications or infrastructural changes have been considered, if relevant

### CODE REVIEW

**Review by jessex — COMMENTED** [review]
@colincadams the code changes look good to me so far! I also did some local testing and the basic toggle functionality appears to be styled and working correctly. However, I think you'll do a better job of testing all of the data-related changes specified in the ticket.

**Comment by colincadams** [comment]
Awesome! Sorry for the delay on this @nasaownsky, here is a link to the annual fixture file: https://drive.google.com/file/d/1F7tCpTxUWEud6qDoPNB2_ZmBG3VLvx0p/view?usp=sharing

This has the same format as the existing data file, but with annual data instead of monthly. I thought that just providing the data in a separate file would be the easiest way to support this, will that work for you?

Additionally a few comments from my first pass at reviewing this:
1. Only the first chart was switched to bars, the rest still show lines. I believe they should all be bars.
    Screen Shot 2021-04-02 at 11 42 26 AM

2. When dropping in the annual data, it changes the time horizon to "5 years" but displays the last 5 months. When expanding to all time it still tries to show a bar for every month, even though there is only data once per year.
    Screen Shot 2021-04-05 at 1 20 37 PM
    Screen Shot 2021-04-05 at 1 20 50 PM

**Review by colincadams — CHANGES_REQUESTED** [review]
requesting changes based on above comment

**Comment by hobuobi** [comment]
Additionally, we noticed that in some cases, we have to use older data than what is available in order to create the yearly metrics shown with the Annual tab. In these cases, we'll want to add a warning tooltip to let the user know that we are artificially restricted the data, and that there is more recent data available in the Monthly tab. 

I updated the main description for #80 with this change, but copying it down below for convenience. 

----

> Data is restricted to older reports in order to match the other reports shown; however, more recent data is available (as of **[DATE YEAR of most recent data for the metric]**). Use the Reporting Frequency switch at the top of the page to see the most recent monthly data. 

Group 71 (3)

As a bonus, the orange link in the design above ("Reporting Frequency switch") should scroll the user back to the page position where the data reporting frequency link can be found.

**Comment by colincadams** [comment]
For that additional tooltip humphrey described, I think the easiest way to implement it would be to compare the data for that state and metric in the monthly and annual files. If there is newer data in the monthly file, then add that note to the annual block.

Do you think that will work? Otherwise we can do it in the backend when producing the files, but didn't want to clutter them any more if we didn't need to.

**Comment by nasaownsky** [comment]
@colincadams ok, I think that'll work & try to make that work tomorrow. For now please check the latest changes and see if there's new potential problems to fix. Thanks!

File: src/utils/generateChartData.js [hunk 1]
```diff
@@ -126,21 +126,22 @@ const generateChartData = (data, metrics, metricLabels = []) => {
   );
   const { startYear, startMonth, endYear, endMonth } = periods;
 
-  let i = startYear * 12 + startMonth;
-  const lastMonth = endYear * 12 + endMonth;
+  let i = annual ? startYear - 5 : startYear * 12 + startMonth;
```
Comment (colincadams) [inline src/utils/generateChartData.js:129]: This bit I don't totally understand. Why do we need to subtract 5 here? Is it to ensure that there are always at least 5 labels to display in the 5 year chart, and if so why don't we need to do something similar for the monthly data?

**Review by colincadams — COMMENTED** [review]
Just checked and it is looking good! Thanks @nasaownsky! Left one implementation question.

I did realize I made a mistake in the fixture file, left monthly populations for CA, so I'm fixing that now and will upload a new version to Google Drive!

File: src/utils/generateChartData.js (hunk 1 above)
Comment (nasaownsky) [reply src/utils/generateChartData.js:129]: For monthly data we calculate the amount of months, and for annual we need this to display dates with empty data, because without this,  graph will show just previous months of last reported year.

**Comment by nasaownsky** [comment]
@colincadams You can check new tooltip on California dashboard!

**Review by colincadams — CHANGES_REQUESTED** [review]
This is looking really good!! No comments from me on the code, but I do think there is one more edge case to handle.

I just pushed up some fake data for KY that tests the case where a state only provides annual data for a metric, so data for that metric won't be in the monthly file but will be in the annual file.

I think we will want to have new tooltip text for this case as well, but similar to the 'restricted' text. Maybe something like:
> Monthly data is not available in a public report from [STATE]; however, annual data is available (as of [DATE YEAR of most recent data for the metric]). Use the Reporting Frequency switch at the top of the page to see the annual data.

@hobuobi to confirm.

**Comment by nasaownsky** [comment]
@colincadams Described case can happen vice versa. So I've made solution that suites for annual and monthly data, but we need more general text for tooltip since the function doesn't know what data we are viewing. I suggest something like:

> Data is not available in a public report from [STATE] in this data aggregation range; however, data is available in another range. Use the Reporting Frequency switch at the top of the page to see the annual data.

**Comment by hobuobi** [comment]
I think the preference would be to have a separate error message for monthly vs annualized data! If that is possible, I would agree with Colin's solution for the case where data is available in the annual view.

> Monthly data is not available in a public report from [STATE]; however, annual data is available (as of [DATE YEAR of most recent data for the metric]). Use the Reporting Frequency switch at the top of the page to see the annual data.

If we have to do something that works for both cases, I would do something like: 

> Data is not available in a public report from [STATE] at this reporting frequency; however, more data is reported [MONTHLY / ANNUALLY (depending on which tab you're on)]. Use the Reporting Frequency switch at the top of the page to see the most recent available data.

**Comment by nasaownsky** [comment]
@colincadams @hobuobi I've added desired solution for that case. Depending on available data there will be one of two options to display: 
Снимок экрана 2021-04-20 в 16 07 07 Снимок экрана 2021-04-20 в 16 08 19

Please check & confirm.

File: src/components/shared/Card/Card.js [hunk 2]
```diff
@@ -38,31 +39,60 @@ const Card = ({
...
+  return (
+    <div
+      className={cn("Card", className, {
+        "Card--population": isPopulation,
+        "Card--not-available": isNotAvailable || isTooStale,
+      })}
+    >
+      <div className="Card__header">
+        <h3 className="Card__title">{title}</h3>
+        {isTooStale && (
```
Comment (colincadams) [inline src/components/shared/Card/Card.js:61]: In some cases, data is too stale to show on the "Monthly" tab, but we still show it on the annual tab. CA is an example of this:

Screen Shot 2021-04-26 at 3 12 18 PM
Screen Shot 2021-04-26 at 3 12 26 PM

This is pretty minor, but I think in this case it would be good to still show the "too stale" text, but then also add the "Use the reporting frequency switch" text so that they know they can see the stale data in the annual view. It is a bit confusing because it might not always be visible in the annual view, but I think it is better than it is currently. @hobuobi thoughts?

**Review by colincadams — APPROVED** [review]
Sorry for the delay in reviewing this @nasaownsky!

In general I think this is looking really good, I am happy with the new tooltip copy.

I did notice one issue with the frequency slider -- on first load it does not show either side as being selected, but the monthly side should be selected. This only repros if I haven't loaded the website in a while, not sure how to get around that, but I have seen it happen twice.

Screen Shot 2021-04-26 at 3 04 26 PM

One other minor comment, but once @hobuobi okays the text I will approve!

**Comment by nasaownsky** [comment]
@colincadams Latest commit should fix slider, but more test only can be performed when we decided to merge `main` with `v2-jails`.

File: src/components/shared/Card/Card.js (hunk 2 above)
Comment (hobuobi) [reply src/components/shared/Card/Card.js:61]: Oh wait, so this is basically a case where both conditions are applicable, right? 
I think your suggestion is a good one, and I'm happy to go with it! In the future maybe we could make these error messages a bit more modular so that we could actually just show both? But not a thing to address right now, I think this is a small enough edge case

**Review by colincadams — APPROVED** [review]
If you want to address the tooStale + use slider text after this, that seems fine with me. I think making the text more modular is probably the right call.

**Comment by nasaownsky** [comment]
@colincadams Let's cover that case in new Issue. @jessex I think we're ready to merge this pr.

**Comment by jessex** [comment]
@nasaownsky just need you to address the merge conflict that popped up from merging the latest PR and then I will merge. Thanks!

**Comment by nasaownsky** [comment]
@jessex You're good to go!
